docs/enhancements/metricsConfig_redesign.md

```diff
@@ -0,0 +1,172 @@
+
+#### clusterDeploymentLabelSelector
+This would be a new feature, of type [LabelSelector](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#LabelSelector), and we'd use the LabelSelector.MatchLabels and/or LabelSelector.MatchExpressions to match the conditions in order to decide if a metric should be reported.
+This encapsulates slightly more advanced filter logic over the existing clusterdeployment labels. In the example above, `hive_foo_counter` and `hive_foo_gauge` metrics will only be reported for the cluster deployments that are labelled with aro-snowflake, not in limited support, and has the metric value > 10m.
+
+### Implementation Details / Notes
+
+- All the options configured for a metric within clusterDeploymentRelatedMetrics, will work in tandem with each other. For ex, if all possible options are specified for a metric, then that metric will be reported with the additional labels as per additionalClusterDeploymentLabels, and will be reported only if it matches the labels and/or expressions as per clusterDeploymentLabelSelector and if the duration to be reported exceeds the minimumDuration
+- This design allows for grouping of metrics for which you would want to define particular customizations. Grouping allows hiveconfig to be parsable, but implementation of this model would require an efficient solution to validate there are no ambiguous entries for a metric, and storing it within the code for parsing purposes.
+- There needs to be clear logic defining which metrics can have `minimumDuration` versus which ones can't. Perhaps if a metric has `seconds` in the name can be the decider.
+- Current configuration of using a shorthand camelCase key to refer to a corresponding metric while configuring hiveconfig.spec.metricsConfig.metricsWithDuration will be removed, and users will have to use the full metric name in the proposed configuration.
+- Presently, the metrics that support hiveconfig.spec.metricsConfig.metricsWithDuration are optional, so they will not be logged unless they're specified in HiveConfig. After this redesign, the metrics that are currently optional will continue to be optional - that behaviour will not change.
+  For example, after the redesign, hive_cluster_deployments_running_transition_seconds will be reported if and only if metricsConfig.clusterDeploymentRelatedMetrics.metrics[hive_cluster_deployments_running_transition_seconds].minimumDuration is not empty, however hive_cluster_deployment_install_job_duration_seconds, which is a non-optional metric that would support minimumDuration, will be reported even if metricsConfig.clusterDeploymentRelatedMetrics.metrics[hive_cluster_deployment_install_job_duration_seconds].minimumDuration is not specified.
```

Comment:
I forgot to specify that optional metrics will continue to remain optional. Hope this explanation isn't confusing, and metrics doc does specify if a metric is optional or not.

Reply:
Well now hold on.

You're saying that metrics that are *currently* not optional simply because we haven't (yet) plumbed them with minimumDuration ==> these will now _support_ minimumDuration, but will differ from metrics that are *currently* optional in that they _will_ be reported if they are not mentioned in the config??

Naw.

This redesign is an opportunity to make the overall picture -- as in, what a brand new user would see -- consistent and sensible. Reporting some metrics by default and not others because we made a seemingly-arbitrary decision -- or worse, for historical reasons growing out of our interactions with one customer (OSD) -- is not a good look. We should decide on one default behavior across the board. And personally, I'm in favor of reporting *nothing* by default. YAGNI principle.

Now the ugly bit is going to be how to handle the deprecation period. Unfortunately, we're going to have to bake in logic like
```
if hiveconfig has the old metrics config section
  then {these metrics} are reported by default but {these metrics} are not
else if hiveconfig has the new metrics config section
  then report all _and only_ metrics that are mentioned in the config, iff their filters say aye.
else if hiveconfig lacks either section
  then, ummm, I guess use the old behavior during the deprecation period? I can be convinced either way here.
else (hiveconfig has both sections)
  panic
```

Reply:
Copyout from DM brain dump:

IIUC currently a *subset* of duration metrics are optional, and the mechanism for opting into or out of them is to specify a minimum duration below which they will not be reported.

Trying to understand this concept from the [doc](https://github.com/openshift/hive/blob/master/docs/hive_metrics.md) is now triggering my pedantic button. "Optional label support" is not the same thing as a metric being optional. A metric being optional implies that it is not reported by default, but must be opted into. Labels are an attribute of a metric, and whether a particular label is included on a reported metric is orthogonal to whether the metric itself is reported or not.

The thing we're discussing (I think) is the default behavior of a duration-based metric that is not mentioned in hiveconfig: do we report it (for all values of the duration) or not? Put another way, is the default minimumDuration zero or infinity? (Which might be a reasonable way to structure the code btw.)

And my position is: I want to be looking at the metrics picture after this redesign as if it were brand new, in isolation. And in such a view, I don't feel it makes sense to have an arbitrary subset of duration metrics that are reported by default, and the rest are not reported by default.

I say "arbitrary" because, while I respect the notion of identifying some metrics that are "critical" or "important" or whatever, IMO that's a user-specific thing. Any decision we make may _mostly_ make sense to _most_ consumers, but there will be a handful where they'll be like, "oh, **this** one is off by default, why??"

Now, I get where we have an issue if we say all duration metrics are off by default, but the rest are on by default. Same kind of seemingly-arbitrary distinction (though this one is at least a little crisper)

I'm thus inclined to say all metrics should be opt-in. This could be a separate section of metricsConfig with a simple list of metric names. (Do we require you to duplicate the cd-related ones in that list and in clusterDeploymentRelatedMetrics? Open to discussion.)

> This has been the trend that I have observed across different services also.

I'm not aware of any other service that has opt-in metrics. Are you?

Reply:
Summary of our discussion from DM:
- Make all metrics opt-in as a part of this enhancement (Does this apply only to clusterDeploymentRelatedMetrics or all metrics?)
- Rename `metricNames` to `metricsToReport`. Within clusterDeploymentRelatedMetrics, only `metricsToReport` will be mandatory, all other customizations are optional.
- Basically, if you're using the new config (aka clusterDeploymentRelatedMetrics), you need to specify all the hive metrics you want reported, hive will not report any metric by default. The list of all metrics to be reported sounds easy enough to collate - say from the dashboard (grafana, amplitude etc) config/deployment files, and from the alerting rules.